pkg/compactdictionary/compactdictionary.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compactdictionary
+
+import (
+	"fmt"
+	stdmath "math"
+
+	"github.com/apache/arrow/go/v16/arrow"
+	"github.com/apache/arrow/go/v16/arrow/array"
+	"github.com/apache/arrow/go/v16/arrow/memory"
+)
+
+type Releasable interface {
+	Release()
+}
+
+// compactDictionary copies only the needed values from the old dictionary to the new dictionary.
+// Once all needed values are copied, it updates the indices referencing those values in their new place.
+func CompactDictionary(mem memory.Allocator, arr *array.Dictionary) (*array.Dictionary, error) {
```

Comment:
no need to move it to its own package, we could already access it within the same package

Reply:
Hmm so I did this because Go complains about importing the "github.com/parca-dev/parca/pkg/query" package (the `compactDictionary` function was originally in the flamegraph_arrow.go file which is essentially a `query` package) in the querier.go file with an error of `import cycle not allowed`.

Open to ideas on how to circumvent this.

Reply:
Oh my bad, I thought the querier.go was in the `query` package. Ignore.